File: app/src/main/java/com/irateam/vkplayer/api/AbstractQuery.java
package com.irateam.vkplayer.api;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public abstract class AbstractQuery<T> implements Query<T> {

    private static final int THREAD_COUNT = 3;
    private static final ExecutorService EXECUTOR_SERVICE =
            Executors.newFixedThreadPool(THREAD_COUNT);
    protected static final Handler UI_HANDLER = new Handler(Looper.getMainLooper());

    private Future<T> task;

    @Override
    public T execute() throws RuntimeException {
        try {
            task = EXECUTOR_SERVICE.submit(this::query);
            return task.get();
        } catch (InterruptedException | ExecutionException e) {
            return null;
        }
    }

    @Override
    public void execute(Callback<T> callback) {
        AsyncCallableAdapter<T> callable = new AsyncCallableAdapter<T>(this, callback);
        task = EXECUTOR_SERVICE.submit(callable);
    }

    @Override
    public void cancel() {
        task.cancel(true);
    }

    protected abstract T query() throws Exception;

    private static class AsyncCallableAdapter<V> implements Callable<V> {

        private final AbstractQuery<V> query;
        private final Callback<V> callback;

        private AsyncCallableAdapter(AbstractQuery<V> query, Callback<V> callback) {
            this.query = query;
            this.callback = callback;
        }

        @Override
        public V call() throws Exception {
            V result = null;
            try {
                result = query.query();
                notifyComplete(result);
            } catch (Exception e) {
                e.printStackTrace();
                notifyError();
            }
            return result;
        }

        private void notifyComplete(V result) {
            if (callback != null) {
                UI_HANDLER.post(() -> callback.onComplete(result));
            }
        }

        private void notifyError() {
            if (callback != null) {
                UI_HANDLER.post(callback::onError);
            }
        }
    }

}

File: app/src/main/java/com/irateam/vkplayer/api/AbstractQuery.kt
package com.irateam.vkplayer.api

import android.os.Handler
import android.os.Looper
import java.util.concurrent.Callable
import java.util.concurrent.ExecutionException
import java.util.concurrent.Executors
import java.util.concurrent.Future

abstract class AbstractQuery<T> : Query<T> {

	private var task: Future<T>? = null

	@Throws(RuntimeException::class)
	override fun execute(): T = try {
		val task = EXECUTOR_SERVICE.submit<T> { this.query() }
		this.task = task
		task.get()
	} catch (e: InterruptedException) {
		throw e;
	} catch (e: ExecutionException) {
		throw e;
	}

	override fun execute(callback: Callback<T>) {
		val callable = AsyncCallableAdapter(this, callback)
		task = EXECUTOR_SERVICE.submit(callable)
	}

	override fun cancel() {
		task?.cancel(true)
	}

	@Throws(Exception::class)
	protected abstract fun query(): T

	private class AsyncCallableAdapter<V> : Callable<V> {

		private val query: AbstractQuery<V>
		private val callback: Callback<V>

		constructor(query: AbstractQuery<V>, callback: Callback<V>) {
			this.query = query
			this.callback = callback
		}

		@Throws(Exception::class)
		override fun call(): V = try {
			val result = query.query();
			notifyComplete(result)
			result
		} catch (e: Exception) {
			e.printStackTrace()
			notifyError()
			throw e;
		}

		private fun notifyComplete(result: V) {
			UI_HANDLER.post { callback.onComplete(result) }
		}

		private fun notifyError() {
			UI_HANDLER.post { callback.onError() }
		}
	}

	companion object {

		private val THREAD_COUNT = 3
		private val EXECUTOR_SERVICE = Executors.newFixedThreadPool(THREAD_COUNT)
		val UI_HANDLER = Handler(Looper.getMainLooper())
	}

}
